### PR TITLE
Allow IPFS Key Export As Mnemonic Phrase

### PR DESCRIPTION
**Pull request by bonedaddy**

### CODE REVIEW

File: keystore.go [hunk 1]
```diff
@@ -83,3 +84,42 @@ func (km *KeystoreManager) CreateAndSaveKey(keyName string, keyType, bits int) (
...
+	}
+	valid := pk.Equals(pk2)
+	if !valid {
+		return "", errors.New("failed to validate key")
+	}
+	return phrase.String(), nil
+}
+
+// MnemonicToKey takes an exported mnemonic phrase, and converts it to a private key
+func (km *KeystoreManager) MnemonicToKey(stringPhrase string) (ci.PrivKey, error) {
```
Comment (bobheadxi) [inline keystore.go]: it doesnt look like this function needs to be a part of `KeystoreManager`

also, `phrase` might be an adequate name for the argument

File: keystore.go [hunk 2]
```diff
@@ -83,3 +84,42 @@ func (km *KeystoreManager) CreateAndSaveKey(keyName string, keyType, bits int) (
...
+	pkBytes, err = mnemonics.FromPhrase(phrase, mnemonics.English)
+	if err != nil {
+		return "", err
+	}
+	pk2, err := ci.UnmarshalPrivateKey(pkBytes)
+	if err != nil {
+		return "", err
+	}
+	valid := pk.Equals(pk2)
+	if !valid {
```
Comment (bobheadxi) [inline keystore.go]: could use the one-line syntax here

```go
if valid := ... ; !valid { ... }
```

File: keystore.go [hunk 3]
```diff
@@ -83,3 +84,42 @@ func (km *KeystoreManager) CreateAndSaveKey(keyName string, keyType, bits int) (
 
 	return pk, nil
 }
+
+// ExportKeyToMnemonic is used to take an IPFS key, and return a human-readable friendly version.
+// The idea is to allow users to easily export the keys they create, allowing them to take control of their records (ipns, tns, etc..)
+func (km *KeystoreManager) ExportKeyToMnemonic(keyName string) (string, error) {
```
Comment (bobheadxi) [inline keystore.go]:
```suggestion
func (km *KeystoreManager) ExportKeyAsMnemonic(keyName string) (string, error) {
```

File: keystore.go [hunk 4]
```diff
@@ -83,3 +84,42 @@ func (km *KeystoreManager) CreateAndSaveKey(keyName string, keyType, bits int) (
...
+	}
+	pkBytes, err := pk.Bytes()
+	if err != nil {
+		return "", err
+	}
+	phrase, err := mnemonics.ToPhrase(pkBytes, mnemonics.English)
+	if err != nil {
+		return "", err
+	}
+	pkBytes, err = mnemonics.FromPhrase(phrase, mnemonics.English)
```
Comment (bobheadxi) [inline keystore.go]: I dont think theres a need to check the phrase is equal to the key - this is something that can be enforced in tests instead

File: keystore_test.go [hunk 5]
```diff
@@ -127,3 +127,46 @@ func TestGetKey(t *testing.T) {
 
 	fmt.Printf("%+v\n", pk2.GetPublic())
 }
+
+func TestKeystoreManager_ExportKey(t *testing.T) {
```
Comment (bobheadxi) [inline keystore_test.go]: this would be a good candidate for a table-driven test, as we might want to check edge cases for key export and add regression cases easily if any issues come up

**Review by bobheadxi — CHANGES_REQUESTED** [review]

File: keystore.go (hunk 1 above)
Comment (bonedaddy) [reply keystore.go]: :man_facepalming: hahahaha yea `phrase` is indeed better than `stringPhrase`

and no it does not need to be a part

File: keystore_test.go [hunk 6]
```diff
@@ -127,3 +127,50 @@ func TestGetKey(t *testing.T) {
...
+	type args struct {
+		name    string
+		keyType int
+		size    int
+	}
+	tests := []struct {
+		name string
+		args args
+	}{
+		{"EDKey-Success", args{"edkey1", ci.Ed25519, 256}},
```
Comment (bobheadxi) [inline keystore_test.go:150]: 😍

**Review by bobheadxi — APPROVED** [review]
clean!
